### DISCUSSION TITLE
Custom Auth Driver

### DISCUSSION THREAD

**Discussion by awesome-developer-20** [directus/directus#11024] [Feature Requests]
Is there a way to add a custom auth driver without changing directus source code here https://github.com/directus/directus/tree/main/api/src/auth/drivers

Maybe I could add my auth driver in extensions just like interfaces and displays

Actually, my client already has users roles database setup for his organization in a weird and unconventional way and they want to use that same data for directus portal login.

Directus is really the best CMS I could ever wish. Thank you for your support ❤️

**Comment by oreilles** [comment]
Thanks for the kind words ❤️
Auth driver extensions are not possible at the moment and are not on the roadmap - consider forking Directus to write your own in the meantime or [sponsoring the feature request](https://v8.docs.directus.io/getting-started/supporting-directus.html#financial-support).

**Comment by awesome-developer-20** [comment]
Thanks for your quick reply.

I keep forking my last option.

Is there a way I could implement it as extension like interfaces/displays?
I'm ready to make a PR but I'll need help how should I implement and organize it.

**Reply by benhaynes** [reply]
CC @aidenfoxx

**Reply by aidenfoxx** [reply]
@benhaynes That is a better question for @rijkvanzanten than me. My focus was mostly on integrating LDAP/OpenID into the existing core of the project.

I think custom backed logic could be done mostly through hooks, and I'm sure it's possible to add some way to manage the frontend though a kind of "interfaces" system.

**Reply by rijkvanzanten** [reply]
There's currently no way to inject auth drivers as extensions. Once we've finalized the ability to have extensions that have both app _and_ API components (like auth drivers), we can look into opening auth drivers as extensions 👍🏻

**Reply by awesome-developer-20** [reply]
> You were already going in the right direction: Try having a look at how `registerEndpoints` / `registerHooks` work and match their behavior. We have unfortunately have plenty to do and cannot support you through this process at the moment.

Should I proceed with @Oreilles approach?

**Reply by awesome-developer-20** [reply]
> There's currently no way to inject auth drivers as extensions. Once we've finalized the ability to have extensions that have both app _and_ API components (like auth drivers), we can look into opening auth drivers as extensions 👍🏻

@rijkvanzanten  So extensions don't include api components? I didn't realized that 😅

**Reply by awesome-developer-20** [reply]
If so, what do you suggest I should do now?

**Reply by rijkvanzanten** [reply]
> > You were already going in the right direction: Try having a look at how `registerEndpoints` / `registerHooks` work and match their behavior. We have unfortunately have plenty to do and cannot support you through this process at the moment.
> 
> Should I proceed with @Oreilles approach?

That'll work, or you could wait for / prioritize by sponsoring the feature request to make auth drivers official extension types 🙂 

> @rijkvanzanten So extensions don't include api components? I didn't realized that 

They do! The only limitation right now is that we have app _or_ API extensions. We don't currently yet have an extension type that has both app "parts" and API "parts" at the same time, which would be the case for auth drivers 👍🏻

**Reply by awesome-developer-20** [reply]
> > @rijkvanzanten So extensions don't include api components? I didn't realized that
> 
> They do! The only limitation right now is that we have app _or_ API extensions. We don't currently yet have an extension type that has both app "parts" and API "parts" at the same time, which would be the case for auth drivers 👍🏻

Ohh now I got your point. I didn't even considered app part. I just thought of auth driver as pure api extension.

Ok no worries. I'll fork directus and will directly modify it. 🙂 

Sorry to disturb you and Thanks a lot for you support ❤️

**Reply by awesome-developer-20** [reply]
> That'll work, or you could wait for / prioritize by sponsoring the feature request to make auth drivers official extension types 🙂

Actually I am still an undergrad student. I'm doing this internships to pay my bills. 😅 

I really support open source development and like directus a lot. I'll definitely support directus once I staring earning my own money 👍

**Reply by awesome-developer-20** [reply]
I'm happy to contribute in directus development. Let me know if you have any tasks that I can do 🙂

**Reply by awesome-developer-20** [reply]
Should I put this in Issues as new feature?
Maybe someday when you finalize on design, I'll start working on this

**Reply by rijkvanzanten** [reply]
> I'm happy to contribute in directus development. Let me know if you have any tasks that I can do 🙂

Anything under Issues that catches your attention 🙂 

> Should I put this in Issues as new feature?

We actually track feature requests right here under _Discussions_, so i'll mark this one as a feature request 👍🏻

**Reply by awesome-developer-20** [reply]
Great Thanks 👍

**Reply by awesome-developer-20** [reply]
@rijkvanzanten just last question

Instead of custom auth driver, Is there a way to use same login page app component and just change api to use custom endpoint for authentication and roles?

Then I can write a simple api that validates the email and password and returns it's roles if email and password are correct.

**Reply by awesome-developer-20** [reply]
Maybe I could add "AUTH_API" env variable which points to any api endpoint and then if "AUTH_API" env variable is set, directus will send the payload there and get the user object instead of getting it from directus user table and provider [here](https://github.com/directus/directus/blob/main/api/src/services/authentication.ts#L60)

**Reply by awesome-developer-20** [reply]
Just found out https://github.com/directus/directus/pull/10375
@rijkvanzanten I think @viters's solution will work for my use case

**Reply by awesome-developer-20** [reply]
I'm totally confused now 😕 

I'll write a separate Q&A for my use case

Update: https://github.com/directus/directus/discussions/11067

**Reply by viters** [reply]
Regarding your use case #11067 and my PR #10375:
My PR allows to overwrite user authentication implementation. This means that you can verify the user in your own way (e.g. using external API or some other database) and create accountability object by yourself. Accountability in Directus specifies the user's ID and role (and some role-derived values, like "hasAppAccess"). Although, in my process, I create a "Directus user" from "my user", to keep everything in Directus working (like relations user_created, user_added).

I have remote auth provider and authorization proxy. Users in my app do not authenticate to Directus, but to my other auth service. Later, the authorization proxy redeems their session for short living JWT, and this JWT is used for authentication in Directus and setting correct role.

Regarding roles, if you have roles somewhere else, you will need counterparts in Directus. Moreover, if you want to manage permissions for Directus collections outside of Directus, my PR will not be enough.

**Reply by awesome-developer-20** [reply]
That's perfect thanks 😊

**Reply by awesome-developer-20** [reply]
@viters
What's the status of your PR?
What are you using as a workaround till this PR merges?

**Reply by viters** [reply]
@awesome-developer-20 
My PR is ready and waits for review, I will have to merge it with Directus master yet after approval.

My current workaround is freezing the Directus accountability before it comes to Diretcus internal middleware. As you cannot freely modify middlewares on your own (reordering or disabling, https://github.com/directus/directus/pull/8714#issuecomment-943465490), you must hack it so it cannot change what you have done. You can achieve it by using `Object.defineProperty`.

```js
Object.defineProperty(req, 'accountability', {
  get() {
    return this._value;
  },
  set(newValue) {
    this._value = {
      ...(typeof newValue === 'object' ? newValue : {}),
      user: user.id,
      role: authenticatedRole.id,
      admin: false,
      app: false,
      ip: req.ip.startsWith('::ffff:') ? req.ip.substring(7) : req.ip,
      userAgent: req.get('user-agent'),
      session: payload.session,
    }
  },
});
```

Thats what I do right now on hook `middlewares.before`.

**Reply by joselcvarela** [reply]
Nice hack 😅

**Reply by awesome-developer-20** [reply]
@viters Thank you so much 🙂

**Comment by awesome-developer-20** [comment]
Maybe I could use a dynamic map [here](https://github.com/directus/directus/blob/main/api/src/auth.ts#L65) and I could add all the auth driver extensions [here](https://github.com/directus/directus/blob/main/api/src/extensions.ts#L173) to the auth drivers map.

**Reply by oreilles** [reply]
That is a great idea. We cannot guarantee nor provide estimate for the time it'll take to review your PR once it's done, but I do believe it's a new feature that many people could find useful, so thank you for contributing 👍

**Reply by awesome-developer-20** [reply]
> We cannot guarantee nor provide estimate for the time it'll take to review your PR once it's done

Yeah no problem, I can understand your situation. I'll use my directus repo till that time.


I just want to know what's the best way of implementing this. For example, whether I should use pinia store or global dynamic map? Do I have to add docs? Do I have to add it in extensions template? Should I use some other approach?

Can someone from your team guide me with all the process?

**Reply by oreilles** [reply]
You were already going in the right direction: Try having a look at how `registerEndpoints` / `registerHooks` work and match their behavior. We have unfortunately have plenty to do and cannot support you through this process at the moment.

**Reply by awesome-developer-20** [reply]
Ok

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
